### DISCUSSION TITLE
how to make a widget be focusable

### DISCUSSION THREAD

**Discussion by pm100** [emilk/egui#4812] [General]
I have a window that I treat as a raw paint canvas. I want to receive kb events (and have the window highlighted to show its accepting keystrokes). I dont expect egui to do anything with the keys, I grab them myself. At the moment I grab them when I see nobody else has focus  (which is a bit lame). I did this
```
        let sense = Sense::focusable_noninteractive();
        let (response, painter) = ui.allocate_painter(draw_area_size, sense);
```
In the draw of the window area (ui is the window's .show ui). But I still dont get the focus

**Comment by rustbasic** [comment]
Please check that it is in the form below.

```
response.request_focus();
```

**Reply by pm100** [reply]
THats not what I need. I dont want to force the focus on that widget. I want it to get focus on tab, click etc  .
I tried Sense::click() too, no difference

**Reply by pm100** [reply]
i also tried

 ctx.memory_mut(|mem| mem.interested_in_focus(self.id));

right at the end of the update logic, nothing

**Comment by YgorSouza** [comment]
Would something like this work? It gets focus when you click on it or tab until you reach it, and it loses focus when you click away or press Escape, like the TextEdit.

```rs
impl eframe::App for MyApp {
    fn update(&mut self, ctx: &egui::Context, _frame: &mut eframe::Frame) {
        egui::CentralPanel::default().show(ctx, |ui| {
            let mut size = ui.available_size();
            size.y /= 3.0;
            focusable_rect(ui, size, Color32::RED);
            focusable_rect(ui, size, Color32::GREEN);
            focusable_rect(ui, size, Color32::BLUE);
        });
    }
}

fn focusable_rect(
    ui: &mut egui::Ui,
    size: impl Into<egui::Vec2>,
    color: egui::Color32,
) -> Response {
    let r = ui.allocate_response(size.into(), Sense::click());
    if r.clicked() {
        r.request_focus();
    }
    if r.has_focus() {
        ui.painter().rect_filled(r.rect, 0.0, color);
    } else {
        ui.painter()
            .rect_filled(r.rect, 0.0, color.gamma_multiply(0.1));
    }
    r
}
```

https://github.com/user-attachments/assets/a622293c-f853-402b-a9b1-03e9fb0d0713

**Reply by pm100** [reply]
That works, the non obvious thing is that the focus does not belong to the main window but the the pseudo widget returned by the allocate_painter call, and I have no control over its ID, Reading the code seems to suggest its semi random (uses auto_id) but in fact its the same random all the time, so I just save the response.id. (I need it to test if I have focus before grabbing the key strokes)

**Reply by YgorSouza** [reply]
The Id is a hash, it stays the same if the UI tree stays the same. You can also force a specific Id with [Ui::push_id](https://docs.rs/egui/0.28.1/egui/struct.Ui.html#method.push_id).
